### PR TITLE
SB-21731: Certificate generator

### PR DESCRIPTION
**Pull request by PradyumnaNagendra**
<table>
<td>
  <a href="https://project-sunbird.atlassian.net/browse/SB-21731" title="SB-21731" target="_blank">StorySB-21731</a>  Migration to Flink Job & Refactoring - Certificate Generation job 
  </td></table>
  <br />
 

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

### Type of change

Please choose appropriate options.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### How Has This Been Tested?

Please describe the tests that you ran to verify your changes in the below checkboxes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Ran Test A
- [ ] Ran Test B

**Test Configuration**:
* Software versions:
* Hardware versions:

### Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by maheshkumargangula — CHANGES_REQUESTED** [review]

**Comment by maheshkumargangula** [comment]
@PradyumnaNagendra - let us not have this big PRs next time. There are some more review feedback.

1. The Rest API calls are not successful, we should through the exception. Check UserFeed and Notifier ProcessFunctions.
2. Mutable variables used at few places. We can change that logic to use immutable.
3. To construct Java Map, we used inline put methods. We can change it and create Scala map and use `asJava` to convert.

File: credential-generator/certificate-processor/src/main/scala/org/sunbird/incredible/pojos/valuator/ExpiryDateValuator.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+            case "d" => cal.add(Calendar.DATE, getDigits(expiry))
+            case "m" => cal.add(Calendar.MONTH, getDigits(expiry))
+            case "y" => cal.add(Calendar.YEAR, getDigits(expiry))
+            case _ => //break
+
+          }
+        }
+        simpleDateFormat.format(cal.getTime)
+      } catch {
+        case e: ParseException => null
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/certificate-processor/src/main/scala/org/sunbird/incredible/pojos/valuator/ExpiryDateValuator.scala]: removed

File: credential-generator/collection-certificate-generator/src/main/scala/org/sunbird/job/certgen/domain/Event.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  def keys: Map[String, String] = readOrDefault[Map[String, String]]("edata.keys", Map[String, String]())
+
+  def logo: String = readOrDefault[String]("edata.logo", "")
+
+  def certificateDescription: String = readOrDefault[String]("edata.description", "")
+
+  def related: Map[String, AnyRef] = readOrDefault[Map[String, AnyRef]]("edata.related", Map[String, AnyRef]())
+
+  def validEvent(): Boolean = {
+    true
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/collection-certificate-generator/src/main/scala/org/sunbird/job/certgen/domain/Event.scala]: this is unused. hence removed

File: credential-generator/collection-certificate-generator/src/main/scala/org/sunbird/job/certgen/task/CertificateGeneratorStreamTask.scala [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      .rebalance
+      .process(new CertificateGeneratorFunction(config, httpUtil, storageService))
+      .name("collection-certificate-generator")
+      .uid("collection-certificate-generator")
+      .setParallelism(config.parallelism)
+
+    processStreamTask.getSideOutput(config.failedEventOutputTag)
+      .addSink(kafkaConnector.kafkaStringSink(config.kafkaFailedEventTopic))
+      .name(config.certificateGeneratorFailedEventProducer)
+      .uid(config.certificateGeneratorFailedEventProducer)
```
[the remark this answers is not in the document]
Comment (PradyumnaNagendra) [reply credential-generator/collection-certificate-generator/src/main/scala/org/sunbird/job/certgen/task/CertificateGeneratorStreamTask.scala]: Removed the failed output tag

**Review by maheshkumargangula — APPROVED** [review]

**Pull request merged**